eval.go

```diff
@@ -450,26 +476,47 @@ func gatherMatches(mt matchTree, known map[matchTree]bool) []*candidateMatch {
 	}
 	cands = res
 
-	// Merge adjacent candidates. This guarantees that the matches
-	// are non-overlapping.
-	sort.Sort((sortByOffsetSlice)(cands))
-	res = cands[:0]
-	for i, c := range cands {
-		if i == 0 {
+	if merge {
+		// Merge adjacent candidates. This guarantees that the matches
+		// are non-overlapping.
+		sort.Sort((sortByOffsetSlice)(cands))
+		res = cands[:0]
+		for i, c := range cands {
+			if i == 0 {
+				res = append(res, c)
+				continue
+			}
+			last := res[len(res)-1]
+			lastEnd := last.byteOffset + last.byteMatchSz
+			end := c.byteOffset + c.byteMatchSz
+			if lastEnd >= c.byteOffset {
+				if end > lastEnd {
+					last.byteMatchSz = end - last.byteOffset
+				}
+				continue
+			}
+
 			res = append(res, c)
-			continue
 		}
-		last := res[len(res)-1]
-		lastEnd := last.byteOffset + last.byteMatchSz
-		end := c.byteOffset + c.byteMatchSz
-		if lastEnd >= c.byteOffset {
-			if end > lastEnd {
-				last.byteMatchSz = end - last.byteOffset
+	} else {
+		// Remove overlapping candidates. This guarantees that the matches
+		// are non-overlapping, but also preserves expected match counts.
+		// TODO do we actually need this guarantee?
```

Comment:
I'm not sure. For example if you did the search query `foobar AND oob` we would gather two matches on `foobarbaz`. Is that useful to show the user? This is obviously a contrived example, but it gets more tricky when you are `AND`ing together regexes/etc.

Reply:
Yeah good example, and that would be surprising from a user perspective. It would be great to find a real life example to help motivate. One benefit of keeping this property is from a fundemental perspective, your result count is bounded by `len(document) / min(query atom)`. This feels like a good thing from a systems perspective / future optimizations.

Reply:
Removed the TODO because there is no reason to change that behavior in this PR.